e2e_playwright/st_audio_input_test.py

```diff
@@ -0,0 +1,138 @@
+    expect(audio_input_elements.nth(0)).to_be_visible()
+    expect(audio_input_elements.nth(1)).to_be_visible()
+    expect(audio_input_elements.nth(2)).to_be_visible()
+
+
+def test_check_top_level_class(app: Page):
+    """Check that the top level class is correctly set."""
+    check_top_level_class(app, "stAudioInput")
+
+
+def test_custom_css_class_via_key(app: Page):
+    """Test that the element can have a custom css class via the key argument."""
+    expect(get_element_by_key(app, "the_audio_input")).to_be_visible()
+
+
+def test_snapshots(themed_app: Page, assert_snapshot: ImageCompareFunction):
+    audio_input_element = themed_app.get_by_test_id("stAudioInput").first
+
+    assert_snapshot(audio_input_element, name="st_audio_input_default")
+
+
+@pytest.mark.only_browser("chromium")
+def test_audio_input_works_in_forms(app: Page):
+    app.context.grant_permissions(["microphone"])
+
+    expect(app.get_by_text("Audio Input in Form: None")).to_be_visible()
+
+    form_audio_input = app.get_by_test_id("stAudioInput").nth(1)
+    form_audio_input.get_by_role("button", name="Record").click()
+    time.sleep(1)
+    form_audio_input.get_by_role("button", name="Stop recording").click()
+
+    expect(app.get_by_text("Audio Input in Form: None")).to_be_visible()
+
+    click_form_button(app, "Submit")
+    wait_for_app_run(app)
+
+    expect(app.get_by_text("Audio Input in Form: None")).not_to_be_visible()
+
+
+@pytest.mark.only_browser("chromium")
+def test_feedback_works_with_fragments(app: Page):
+    app.context.grant_permissions(["microphone"])
+
+    expect(app.get_by_text("Runs: 1")).to_be_visible()
+    expect(app.get_by_text("Audio Input in Fragment: None")).to_be_visible()
+
+    fragment_audio_input = app.get_by_test_id("stAudioInput").nth(2)
+    fragment_audio_input.get_by_role("button", name="Record").click()
+    time.sleep(1)
+    fragment_audio_input.get_by_role("button", name="Stop recording").click()
+    wait_for_app_run(app)
+
+    expect(app.get_by_text("Audio Input in Fragment: None")).not_to_be_visible()
+    expect(app.get_by_text("Runs: 1")).to_be_visible()
+
+    fragment_audio_input.get_by_role("button", name="Clear recording").click()
+    wait_for_app_run(app)
+
+    expect(app.get_by_text("Runs: 1")).to_be_visible()
+
+
+@pytest.mark.only_browser("chromium")
+def test_audio_input_basic_flow(app: Page):
+    app.context.grant_permissions(["microphone"])
+
+    audio_input = app.get_by_test_id("stAudioInput").first
+
+    expect(
+        app.get_by_text("This app would like to use your microphone.").first
+    ).not_to_be_visible()
+
+    record_button = app.get_by_role("button", name="Record").first
+    clock = audio_input.get_by_test_id("stAudioInputWaveformTimeCode")
+
+    expect(clock).to_have_text("00:00")
+
+    record_button.click()
+
+    stop_button = audio_input.get_by_role("button", name="Stop recording").first
+    expect(stop_button).to_be_visible()
+
+    time.sleep(2)
+
+    stop_button.click()
```

Comment:
Maybe also add a `wait_for_app_run` here and check via the written value that the value was actually received by the frontend as done in the form and fragment case.